### PR TITLE
Windows: utilize ShellExecuteW

### LINKED ISSUE #16: Stop using CMD with Windows
The current Windows incarnation is brittle:

https://github.com/pkg/browser/blob/0a3d74bf9ce488f035cf5bc36f753a711bc74334/browser_windows.go#L9-L12

it survives ampersands, but not other characters:

~~~go
package main
import "github.com/pkg/browser"
func main() {
   browser.OpenURL("http://example.com?aa bb&cc")
}
~~~

better would be to call RunDLL32, no replacer needed. Examples:

~~~sh
rundll32 url,OpenURL "http://example.com?aa bb&cc"
rundll32 url,FileProtocolHandler "http://example.com?aa bb&cc"
~~~

https://github.com/skratchdot/open-golang/blob/master/open/exec_windows.go

Or call browser directly via the BROWSER environment variable. This is what
Python does:

~~~py
import webbrowser
webbrowser.open('http://example.com?aa bb&cc')
~~~

https://github.com/python/cpython/blob/master/Lib/webbrowser.py

Or better still would be to call `ShellExecuteW` on Windows:

https://docs.microsoft.com/windows/win32/api/shellapi/nf-shellapi-shellexecutew

This is what D does:

~~~d
import std.process;
void main() {
   browse("http://example.com?aa bb&cc");
}
~~~

https://github.com/dlang/phobos/blob/master/std/process.d

and Nim:

~~~nim
import browsers
openDefaultBrowser("http://example.com?aa bb&cc")
~~~

https://github.com/nim-lang/Nim/blob/devel/lib/pure/browsers.nim

### PR DESCRIPTION
**Pull request by 89z**
On Windows, use the `ShellExecuteW` syscall instead of an external command. The
`HideWindow` option is no longer configurable, as `setFlags` assumes an external
command is to be run. Long term, would be better to refactor the package, as
different Window behavior can be configured via the last argument to
`ShellExecuteW` (`nShowCmd`).

https://docs.microsoft.com/windows/win32/api/shellapi/nf-shellapi-shellexecutew

Fixes #16

### CODE REVIEW

File: browser_windows.go
Comment (davecheney) [inline browser_windows.go:6]: please move this to line 8-9, this constant so the constant isn't exported by the package.

File: browser_windows.go
Comment (davecheney) [inline browser_windows.go:9]: Is this. available in the x/sys package? If not, would they accept this function so we don't have to own it.

**Review by davecheney — CHANGES_REQUESTED** [review]
Thank you for this change.

I don't have any way to test this, but it looks like it should work.

Modulo my comments, if another windows user can 👍  or 👎  this change then we can merge it.

File: browser_windows.go
Comment (89z) [reply browser_windows.go:9]: It is available:

https://pkg.go.dev/golang.org/x/sys/windows#ShellExecute

but personally I dont like the signature. They went with `*uint16`, when `string` is the closer match, and with the `mkwinsyscall` tool, really no reason to have to put the conversion on the end user.

Comment (davecheney) [reply browser_windows.go:9]: I would prefer to use the version x/sys and to work with the maintainers of that package to adjust the signature if they agree.

Comment (89z) [reply browser_windows.go:9]: Understand. I did a PR on `x/sys` for a similar issue https://github.com/golang/sys/pull/96 I think they are on code freeze until February though. I was going to wait to see how that panned out before taking further action. Regarding this repo, as a stopgap I can just switch it to the package version if thats preferable.

Comment (davecheney) [reply browser_windows.go:9]: Thanks for looking at that. I’ll merge this as is and let’s come back when 1.16 is out and see how much can be up streamed into x/sys

**Comment by pasamio** [comment]
The dependency on x/sys/windows seems to break cross compilation with Windows as a target. Is this intended?

**Comment by davecheney** [comment]
@pasamio oh dear. I'll revert this.

**Comment by 89z** [comment]
https://github.com/pkg/browser/pull/28

**Comment by 89z** [comment]
the comment from @pasamio isnt really helpful, unless they can specific the host system.

**Comment by pasamio** [comment]
The build host is a Linux AMD64 Docker container for Golang 1.14.10, the standard Buster one available from Docker. It's a little dated now but not too old. 

go build is called with `CGO=0 GOOS=windows` and the error reported was:

```
/go/src/github.com/pkg/browser/zbrowser_windows.go:9:2: cannot find package "golang.org/x/sys/windows" in any of:
	/usr/local/go/src/golang.org/x/sys/windows (from $GOROOT)
	/go/src/golang.org/x/sys/windows (from $GOPATH)
```

I didn't intend this as an immediate revert, I wanted to pose the question of if this is intended. Perhaps I should have phrased it as expected.

**Comment by 89z** [comment]
Using this file:

~~~
package main
import _ "golang.org/x/sys/windows"

func main() {
}
~~~

and this invocation:

~~~
go mod init test
go build
~~~

I get a successful build with:

~~~
go version go1.15.3 windows/amd64
~~~

I cant speak to any cross compilation issues, Ive never done any cross
compilation with Go, only C. I would say either something needs to be added in
the way of a Go build contraint, or maybe the Docker instance needs to be
updated. I dont know, but I know for sure its valid code in a native Windows
environment.

**Comment by pasamio** [comment]
Initially I tried this and it resulted in an error:

```
$ go get -u golang.org/x/sys/windows
package golang.org/x/sys/windows: build constraints exclude all Go files in /go/src/golang.org/x/sys/windows
```

And that failed the build in the environment because it's an error code response. 

Running it interactively in the Docker environment I noticed it's still pulling the files and when run with `GOOS=windows` set, this seems to not flag the result as an error properly. Not what I expected.

I think I reverted back to the earlier commit/pre-revert and tried a build which appeared to work. Apologise for the noise.

**Comment by davecheney** [comment]
Thanks for confirming. Please raise an issue if there are further difficulties

**Pull request merged**
